nixos/modules/services/mail/dovecot.nix

```
@@ -541,18 +636,17 @@ in
         perProtocol.imap.enable = [ "imap_quota" ];
       };
 
-      sieve.plugins =
-        optional (cfg.imapsieve.mailbox != []) "sieve_imapsieve"
-        ++ optional (cfg.sieve.pipeBins != []) "sieve_extprograms";
-
-      sieve.globalExtensions = optional (cfg.sieve.pipeBins != []) "vnd.dovecot.pipe";
-
-      pluginSettings = lib.mapAttrs (n: lib.mkDefault) ({
-        sieve_plugins = concatStringsSep " " cfg.sieve.plugins;
-        sieve_extensions = concatStringsSep " " (map (el: "+${el}") cfg.sieve.extensions);
-        sieve_global_extensions = concatStringsSep " " (map (el: "+${el}") cfg.sieve.globalExtensions);
-        sieve_pipe_bin_dir = sievePipeBinScriptDirectory;
-      } // sieveScriptSettings // imapSieveMailboxSettings);
+      pluginSettings = (lib.mapAttrs (n: lib.mkDefault) (
+        sieveScriptSettings // imapSieveMailboxSettings
+      )) // (lib.mapAttrs (n: lib.mkBefore) (filterAttrs (n: v: v != []) {
+        sieve_global_extensions = (
+          optional (cfg.sieve.pipeBins != []) "+vnd.dovecot.pipe"
+        );
+        sieve_plugins = (
+          optional (cfg.imapsieve.mailbox != []) "sieve_imapsieve"
+          ++ optional (cfg.sieve.pipeBins != []) "sieve_extprograms"
+        );
+      }));
```

```suggestion
      pluginSettings = mkMerge [
        sieveScriptSettings
        imapSieveMailboxSettings
        (mkIf (cfg.sieve.pipeBins != []) {
          sieve_plugins = [ "sieve_extprograms" ];
          sieve_global_extensions =  [ "+vnd.dovecot.pipe" ];
        })
        (mkIf (cfg.imapsieve.mailbox != []) {
          sieve_plugins = [ "sieve_imapsieve" ];
        })
      ];
```
This is getting a bit messy. Also that mkDefault shouldn't be needed anymore since both sieveScriptSettings and imapSieveMailboxSettings will be empty until the user sets the corresponding options (and conflicts should error for these).

---

This is a cool trick! :+1: